src/grain/createGrainAllocation.js

```diff
@@ -0,0 +1,255 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating allocation amounts for
+ * contributors. An allocation contains "receipts" showing how much Grain each
+ * contributor will receive in a distribution, and the "strategy"
+ * used to allocate grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which allocates a fixed budget of grain based on the cred scores
+ * in the most recent completed time interval
+ * - BALANCED, which allocates a fixed budget of grain based on cred scores
+ * across all time, prioritizing paying people who were under-paid historically
+ * (i.e. their lifetime earnings are lower than we would expect given their
+ * current cred score)
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {
+  type Grain,
+  multiplyFloat,
+  ZERO,
+  toFloatRatio,
+  format,
+  DECIMAL_PRECISION,
+} from "./grain";
+
+export const GRAIN_ALLOCATION_VERSION_1 = 1;
+
+export type AllocationStrategy = ImmediateV1 | BalancedV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+|};
+
+export type BalancedV1 = {|
+  +type: "BALANCED",
+  +version: number,
+|};
+
+export type GrainReceipt = {|
+  +address: NodeAddressT,
+  +amount: Grain,
+|};
+
+export type GrainAllocationV1 = {|
+  +version: number,
+  +strategy: AllocationStrategy,
+  +budget: Grain,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Allocation given:
+ * - the strategy we're using
+ * - the amount of Grain to allocate
+ * - the full cred history for all users
+ * - a Map of the total lifetime grain that
+ *   has been distributed to each user
+ */
+export function createGrainAllocation(
+  strategy: AllocationStrategy,
+  budget: Grain,
+  credHistory: CredHistory,
+  lifetimeEarningsMap: Map<NodeAddressT, Grain>
```

Comment:
pastAllocationsMap sounds like it should be a map of nodeAddress > Allocations instead of nodeAddress > Grain. 

How about `lifetimeGrainReceived`?

Reply:
Grain received has the same issue. Why wouldn't a transfer be received grain?

I'm not too worried about how the plurals are interpreted, because it says immediately besides the name it's typed `Map<NodeAddressT, Grain>` not `Map<NodeAddressT, Grain[]>`.

But "received" `Grain` and "earned" `Grain` could be any source of Grain. I think the main context the variable name should add is that it's `Grain` from allocations specifically. Otherwise, the only way to know about this distinction is by already being familiar with the algorithm used here or a comment elsewhere.

Reply:
If we want to specify its from Allocations only, I like "lifetimeGrainAllocation" (i.e. the total amount of grain that was allocated to you) more than "pastAllocationsMap". Will change it to that, unless you have some strong feelings against "lifetimeGrainAllocation"

Reply:
Seems good to me.